### PR TITLE
infoschema: add WithRefillOption for TableByName and TableByID

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: ref #50959 

Problem Summary:

### What changed and how does it work?

Use context to pass the refill option for TableByName/TableByID

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/infoschema/infoschema_v2.go [hunk 1]
```diff
@@ -1456,3 +1465,14 @@ func (is *infoschemaV2) ListTablesWithSpecialAttribute(filter specialAttributeFi
 	}
 	return ret
 }
+
+type refillOption string
```
Comment (lance6716) [inline pkg/infoschema/infoschema_v2.go:1469]:
```suggestion
type refillOption struct{}
```

no need to assign the "refillOptionKey" for it? We don't check the key string.

**Review by lance6716 — APPROVED** [review]

File: pkg/infoschema/infoschema_nokit_test.go [hunk 2]
```diff
@@ -21,6 +21,13 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// GetCache is exported for testing.
+func (is *infoschemaV2) GetCache(tableID int64, schemaVersion int64) bool {
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_nokit_test.go:25]:
```suggestion
func (is *infoschemaV2) HasCache(tableID int64, schemaVersion int64) bool {
```

File: pkg/ddl/foreign_key.go [hunk 3]
```diff
@@ -161,8 +161,9 @@ func checkTableForeignKeysValid(sctx sessionctx.Context, is infoschema.InfoSchem
 	}
 
 	referredFKInfos := is.GetTableReferredForeignKeys(schema, tbInfo.Name.L)
+	ctx := infoschema.WithRefillOption(context.Background(), false)
```
Comment (D3Hunter) [inline pkg/ddl/foreign_key.go:164]: seems it's always false right now

File: pkg/infoschema/infoschema_v2.go [hunk 4]
```diff
@@ -1456,3 +1465,14 @@ func (is *infoschemaV2) ListTablesWithSpecialAttribute(filter specialAttributeFi
...
+
+const refillOptionKey refillOption = "refillOptionKey"
+
+// WithRefillOption controls the infoschema v2 cache refill operation.
+// By default, TableByID does not refill schema cache, and TableByName does.
+// The behavior can be changed by providing the context.Context.
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:1475]: why we have such a difference?

Comment (D3Hunter) [reply pkg/infoschema/infoschema_v2.go:1475]: please add this comment to `TableByName` and TableByID

File: pkg/ddl/foreign_key.go (hunk 3 above)
Comment (tiancaiamao) [reply pkg/ddl/foreign_key.go:164]: Yes, we do not fill cache for TableByName in some place like in a loop that may visit a lot tables.

**Review by tangenta — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**